### PR TITLE
feat: Object Builder - relations & computations

### PR DESCRIPTION
**Pull request by frondeus**

### CODE REVIEW

File: crates/dto/src/materialization.rs [hunk 1]
```diff
@@ -507,6 +423,40 @@ impl<'de> Deserialize<'de> for FieldType {
     }
 }
 
+#[derive(Clone, Debug, PartialEq)]
+// Todo: probably unify it with api::types::view::queried::FullView
```
Comment (Kononnable) [inline crates/dto/src/materialization.rs:427]: TODO

File: crates/dto/src/materialization.rs [hunk 2]
```diff
@@ -237,7 +181,7 @@ pub struct ComputedFilter {
 impl ComputedFilter {
     fn from_rpc(rpc: rpc::schema_registry::ComputedFilter) -> RequestResult<Self> {
         Ok(Self {
-            computation: Computation::from_rpc(&rpc.computation)?,
+            computation: Computation::from_rpc(rpc.computation)?,
```
Comment (Kononnable) [inline crates/dto/src/materialization.rs:184]: All those from_rpc, to_rpc looks very cumber stone? Can we replace them somehow(not necessary in this PR)?
Proc macro is still an option if no out of the box solution exist.

File: crates/materializer-general/src/lib.rs [hunk 3]
```diff
@@ -33,7 +33,7 @@ impl IntoSerialize<MaterializationNotification> for MaterializedView {
                 .rows
                 .into_iter()
                 .map(|row| MaterializationRow {
-                    object_id: row.object_id,
+                    object_ids: row.object_ids,
```
Comment (Kononnable) [inline crates/materializer-general/src/lib.rs:36]: If we need such conversion maybe we could implement `From`? Just so there is one implementation of this conversion in the future, not simple object creation in multiple places.

File: crates/materializer-general/src/plugins/postgres.rs [hunk 4]
```diff
@@ -48,7 +48,11 @@ impl TryFrom<MaterializedView> for PsqlView {
             .rows
             .into_iter()
             .map(|row| {
-                let object_id: Uuid = row.object_id.parse()?;
+                let object_ids = row
+                    .object_ids
+                    .into_iter()
+                    .map(|oid| oid.parse().map_err(anyhow::Error::from))
+                    .collect::<anyhow::Result<_>>()?;
```
Comment (Kononnable) [inline crates/materializer-general/src/plugins/postgres.rs:55]: We're using `anyhow::Result`, as function outcome `Result<Self,anyhow::Error>` - can be simplified(cannot make a code suggestion to line which was not changed in this PR).

File: crates/object-builder/src/buffer_stream.rs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+mod buffer;
+
+use crate::{ObjectIdPair, RowSource};
+use buffer::ObjectBuffer;
+
+pub struct ObjectBufferedStream<S> {
+    buffer: ObjectBuffer,
+    input: S,
+}
```
Comment (Kononnable) [inline crates/object-builder/src/buffer_stream.rs:15]:
```suggestion
pub struct ObjectBufferedStream<S>
where
    S: Stream<Item = Result<(ObjectIdPair, Value)>> + Unpin,
{
    buffer: ObjectBuffer,
    input: S,
}
```
Creation of this object where S is not an Stream doesn't make much logic sense.

File: crates/object-builder/src/buffer_stream.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::{pin_mut, FutureExt, StreamExt};
+
+    // let edges = r#"{
+    //     "1": { "objects": [] }
+    // }"#;
```
Comment (Kononnable) [inline crates/object-builder/src/buffer_stream.rs:60]:
```suggestion
```

File: crates/object-builder/src/buffer_stream/buffer.rs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+impl ObjectBuffer {
+    pub fn new(edges: &HashMap<NonZeroU8, TreeResponse>) -> Self {
+        let mut unfinished_rows: Vec<_> = Default::default();
+
+        let missing = edges
+            .values()
+            .flat_map(|res| res.objects.iter())
+            .map(|obj| (retrieve_all_objects(obj), obj))
+            .map(|(res, obj)| {
+                unfinished_rows.push(Some(UnfinishedRow {
```
Comment (Kononnable) [inline crates/object-builder/src/buffer_stream/buffer.rs:45]: mutating another structure in a `map` looks very weird

File: crates/object-builder/src/pool.rs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+use rpc::edge_registry::edge_registry_client::EdgeRegistryClient;
+use rpc::schema_registry::schema_registry_client::SchemaRegistryClient;
+use tonic::transport::Channel;
+
+pub type SchemaRegistryPool = Pool<SchemaRegistryConnectionManager>;
+pub type SchemaRegistryConn = SchemaRegistryClient<Channel>;
+
+pub type EdgeRegistryPool = Pool<EdgeRegistryConnectionManager>;
+pub type EdgeRegistryConn = EdgeRegistryClient<Channel>;
+#[derive(Clone)]
```
Comment (Kononnable) [inline crates/object-builder/src/pool.rs:11]:
```suggestion

#[derive(Clone)]
```

File: crates/object-builder/src/row_builder.rs [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        let fields = self
+            .view
+            .fields
+            .iter()
+            .map(|(field_def_key, field_def)| {
+                Ok((
+                    field_def_key.into(),
+                    match field_def {
+                        Simple { field_name, .. } => {
+                            //TODO: Use field_type
```
Comment (Kononnable) [inline crates/object-builder/src/row_builder.rs:87]: TODO

File: crates/object-builder/src/row_builder.rs [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+                            let value = object.get(field_name).with_context(|| {
+                                format!(
+                                    "Object ({}) does not have a field named `{}`",
+                                    object_id, field_name
+                                )
+                            })?;
+                            value.clone()
+                        }
+                        Computed { .. } => {
+                            // TODO: In theory we could enable subset of it later
```
Comment (Kononnable) [inline crates/object-builder/src/row_builder.rs:97]: Not just in theory :)

File: crates/object-builder/src/row_builder/utils.rs [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+                            )
+                        })?;
+
+                    Ok(ObjectIdPair {
+                        schema_id: tree_object.relation.parent_schema_id,
+                        object_id: tree_object.object_id,
+                    })
+                }
+                SearchFor::Children => {
+                    // TODO: Send error to materializer so it can do transaction rollup
```
Comment (Kononnable) [inline crates/object-builder/src/row_builder/utils.rs:86]:
```suggestion
                    // TODO: Send error to materializer so it can do transaction rollup #543 
```

**Review by Kononnable — COMMENTED** [review]
Draft stage, but as discussed no major changes are planned.

File: crates/dto/src/edges.rs [hunk 12]
```diff
@@ -0,0 +1,65 @@
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+use crate::{RequestError, RequestResult};
+
+#[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub struct TreeResponse {
```
[the remark this answers is not in the document]
Comment (frondeus) [reply crates/dto/src/edges.rs:7]: Okay, but I'm not the author of this structure. The idea was to have the same name as its in the edge_registry. Would you like to change that as well? If so I'd suggest creating another issue and PR for that

File: crates/dto/src/materialization.rs [hunk 13]
```diff
@@ -1,3 +1,5 @@
+// TODO: Refactor: Split me into smaller modules
```
[the remark this answers is not in the document]
Comment (frondeus) [reply crates/dto/src/materialization.rs:1]: Yes, but its TODO for the future

File: crates/dto/src/materialization.rs (hunk 1 above)
Comment (frondeus) [reply crates/dto/src/materialization.rs:427]: Yes, but its TODO for the future

File: crates/dto/src/materialization.rs (hunk 2 above)
Comment (frondeus) [reply crates/dto/src/materialization.rs:184]: I can write a macro sure, just please, not in this PR

File: crates/object-builder/src/buffer_stream/buffer.rs (hunk 7 above)
Comment (frondeus) [reply crates/object-builder/src/buffer_stream/buffer.rs:45]: Well if you can suggest another, more elegant solution that is as fast or faster than this one, be my guest. I spent a lot of time trying to polish this one

File: crates/object-builder/src/row_builder.rs (hunk 9 above)
Comment (frondeus) [reply crates/object-builder/src/row_builder.rs:87]: There is another issue which covers that. Therefore TODO for another PR

File: crates/materializer-general/src/plugins/postgres.rs (hunk 4 above)
Comment (frondeus) [reply crates/materializer-general/src/plugins/postgres.rs]: Simplified

File: crates/object-builder/src/buffer_stream.rs (hunk 5 above)
Comment (frondeus) [reply crates/object-builder/src/buffer_stream.rs:15]: While yes, can you create this object without using `::new()` constructor?

File: crates/dto/src/edges.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Kononnable) [reply crates/dto/src/edges.rs:7]: I'm fine with another PR - just noticed it during code review.

File: crates/object-builder/tests/README.md [hunk 14]
```diff
@@ -0,0 +1,10 @@
+
+To fix tests by accepting new version:
+
+``` sh
+patch --ignore-whitespace < *.new
+```
+
+TODO:
+* [ ] Edge cases when View is invalid
```
Comment (frondeus) [inline crates/object-builder/tests/README.md:9]: Although these TODO are not done yet, these are mostly for testing not for production code.

**Review by Kononnable — APPROVED** [review]

File: crates/object-builder/.gitignore [hunk 15]
```diff
@@ -0,0 +1,7 @@
+.DS_Store
```
[the remark this answers is not in the document]
Comment (frondeus) [reply crates/object-builder/.gitignore]: removed ds store but about tests ignores - I could move it to the root gitignore, but honestly I don't see that file as an issue. Even if doker copies it, it's just a few bytes.

File: crates/object-builder/tests/README.md (hunk 14 above)
Comment (frondeus) [reply crates/object-builder/tests/README.md:17]: Second TODO is ready, and the first one - well id suggest that we add it in another issue/PR.

I'm gonna create all the issues right now.

File: crates/dto/src/edges.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (frondeus) [reply crates/dto/src/edges.rs:7]: #559

File: crates/dto/src/materialization.rs (hunk 2 above)
Comment (frondeus) [reply crates/dto/src/materialization.rs:184]: #560

File: crates/object-builder/tests/README.md (hunk 14 above)
Comment (frondeus) [reply crates/object-builder/tests/README.md:17]: #561

**Review by Kononnable — APPROVED** [review]
